### ISSUE TITLE
allOf rendered in an unexpected way

### ISSUE THREAD

**Issue by jamescrowley** [asyncapi/html-template#11]
#### Describe the bug

I'm not sure if this is a bug or not, but for me, allOf is rendered in an unexpected way when used in the message payload. Can you confirm if this is deliberate or not? If not, I'll look to fix.

#### How to Reproduce

```
asyncapi: 2.0.0
info:
  title: MQTT API
  version: '2.0.0'

defaultContentType: application/json

channels:
  /v2/abc/data/pack_data:
    description: Publish logging data from devices
    publish:
      operationId: publishPackData
      message:
        name: pack_data
        payload:
          $ref: "#/components/schemas/pack_data"

components:
  schemas:
    pack_data:
      allOf:
      - $ref: "#/components/schemas/common_payload"
      - type: object
        properties:
          cycle_id:
            type: string
    common_payload:
      type: object
      properties:
        device_id:
            type: string
        site_id:
            type: string
```

<img width="789" alt="Screenshot 2020-04-23 at 13 47 24" src="https://user-images.githubusercontent.com/509533/80057011-f8791800-8568-11ea-925b-f341334cc078.png">
#### Expected behavior

I'd expect the use of 'allOf' to be transparent in the documentation (ie as if everything had just been defined explicitly in-line).

**Comment by derberg** [comment]
Makes sense, this is the part of the template responsible for schema -> https://github.com/asyncapi/html-template/blob/master/partials/schema-prop.html

**Comment by jamescrowley** [comment]
This may not have the use case I expected, due to the fact that allOf requires the message to validate the schema of each (as opposed to extend them), and as such you may need to understand the individual elements as a consumer.

See https://stackoverflow.com/questions/22689900/json-schema-allof-with-additionalproperties. I'll investigate this further.

**Comment by jamescrowley** [comment]
Some potential code here: https://github.com/jamescrowley/html-template/tree/flattenAllOf2

**Comment by InTheCloudDan** [comment]
Just adding a 👍 here from conversation on Slack. I'm looking for this feature also. My current usage is pretty basic though and I haven't done any interacting with `allOf` and `additionalProperties` so not sure on the best approach to handling them.

**Comment by derberg** [comment]
@jamescrowley yeap, looking at https://json-schema.org/understanding-json-schema/reference/combining.html#id5 looks like there is no other option but flatten only if `"additionalProperties": false` is not present. Pretty complicated case as maybe there are other combinations too.

I checked quickly with OpenAPI and the swagger UI is flattening always

**Comment by InTheCloudDan** [comment]
Any update on this one? It would be useful feature to me.

**Comment by derberg** [comment]
@InTheCloudDan have a look on the thread in this PR https://github.com/asyncapi/html-template/pull/16 from @jamescrowley. We are pretty much stuck there, appriciate your input into discussion.

**Comment by WaleedAshraf** [comment]
I also have this problem. But for now, we are okay with two schemas in the view. In our case, first schema is always the same for all the messages, So, it's better hidden.
Mostly we are only interested in seeing the second part.

I think best solution will be if on top somewhere, we can provide checkboxes
- [ ] merge allOf schemas
- [ ] expand all schemas
- [ ] collapse all schemas

**Comment by FabienSailliet** [comment]
Has there been any progress on this issue? Would be really useful for me too

**Comment by derberg** [comment]
@FabienSailliet Hi, as you can see there was a PR for that https://github.com/asyncapi/generator/issues/115 but because of concerns mentioned in this issue and the PR, we did not come with an final conclusion how this shoudl be solved. What is your opinion on the topic?

**Comment by FabienSailliet** [comment]
It is to be noted that in the [JSON Schema specification](https://json-schema.org/understanding-json-schema/reference/combining.html#id5), they are stating that `allOf` is not to be used to extend schemas:

>  While it might be surprising, allOf can not be used to “extend” a schema to add more details to it in the sense of object-oriented inheritance.

They later say:

>  There are some proposals to address this in the next version of the JSON schema specification.

But they do not give more details about it, so maybe it could be better to wait for something on their side, I did not check if there were some discussions about it.

In my opinion, I think that if a property is described in multiple objects to be merged, the resulting merged object should contain the most specific given definition, and we should be allowed to do this only if the definitions are compatible (we cannot merge a property that should be an array on one side and an integer on another). By doing it this way, we can have objects allowing additional properties or not, as long as they are not incompatible.

**Comment by WaleedAshraf** [comment]
"[unevaluatedProperties](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-02#section-9.3.2.4)" in draft-2019 solves the problem of using "additionalProperties" with "allOf".

Example,
```json
{
  "$schema": "http://json-schema.org/draft/2019-09/schema",
  "allOf": [
    {
      "type": "object",
      "properties": {
        "veggieName": {
          "type": "string",
        }
      }
    },
    {
      "type": "object",
      "properties": {
        "data": {
          "type": "object",
          "additionalProperties": false,
          "properties": {
            "veggieLike": {
              "type": "boolean",
            }
          }
        }
      }
    }
  ],
  "unevaluatedProperties": false,
}
```

Adding any additional prop to any of `allOf` schema will fail the validation.

**Comment by elnipa** [comment]
Any update on this one? A solution would be highly appreciated to have the same behavior as in Swagger/Redoc for our openAPI rest API.

**Comment by derberg** [comment]
@elnipa could you show some examples? I'm not very familiar with how these tools handle it and what you refer to exactly. Thanks 🙇🏼

**Comment by elnipa** [comment]
> @elnipa could you show some examples? I'm not very familiar with how these tools handle it and what you refer to exactly. Thanks 🙇🏼

Sure, in Redoc the attributes within allOf are flattened and only oneOf attributes are splitted.
Compare for the following example:

```
  exampleChannel:
    description: example channel
    publish:
      message:
        payload:
          allOf:
            - type: object
              properties:
                foo:
                  type: string
            - type: object
              properties:
                bar:
                  type: string
            - oneOf:
                - type: object
                  properties:
                    option1:
                      type: boolean
                - type: object
                  properties:
                    option2:
                      type: integer
```

## asyncAPI
![image](https://user-images.githubusercontent.com/34566347/101666242-3e2e1680-3a4e-11eb-97b1-8c6e7e9a0b7b.png)

## openAPI (Redoc)
![image](https://user-images.githubusercontent.com/34566347/101666293-5140e680-3a4e-11eb-8ffc-ff403e6be5ad.png)
![image](https://user-images.githubusercontent.com/34566347/101666314-59008b00-3a4e-11eb-95ac-e2c2140d5ff6.png)

**Comment by derberg** [comment]
@elnipa thanks a lot, it helps!

**Comment by Z5eyhS0uubejR0SVmX2O** [comment]
I second the alignment with redoc which @elnipa is proposing.

**Comment by derberg** [comment]
anyone wants to give this one a try again? I guess some work from @jamescrowley could be reused

**Comment by Z5eyhS0uubejR0SVmX2O** [comment]
Going along with this, it would be super nice if the `title` of the object in the OneOf/AllOf rendered as Redoc does. I updated my examples to include this.

**Comment by Z5eyhS0uubejR0SVmX2O** [comment]
Identified further concern with how properties in addition to those defined in OneOf/AllOf are rendered, added to the examples I link to (just so I can provide a URL that shows how redoc-cli renders it).

**Comment by stsydow** [comment]
I just want to note that the issue still exists.
May be it has to do with specs split into different files?

**Comment by magicmatatjahu** [comment]
I don't know if we should handle that issue in the near future. `allOf` in the JSON Schema doesn't mean that "objects" should merged but the payload itself should be compatible with all schemas defined under `allOf` and this is a big difference, because during merging we may forget some elements and the final schema will not be correct referring to the original. WDYT @derberg @fmvilas? Of course `oneOf` we should try to visualise in better way (we have separate issue for that - https://github.com/asyncapi/asyncapi-react/issues/278), but that problem is only related to the `allOf`

**Comment by derberg** [comment]
I'm supporting merging only if this is optional as I stated above because there might be people that designed their schemas in a way that it will always be valid when we merge in one.

**Comment by derberg** [comment]
I'm wondering if we should not have it solved on parser side, have a helper that allows to return merged schema, so in react template we just make it optional by props.

anyway moving to react as it should be solved there
